lib/libvfio-user.c

```diff
@@ -462,186 +462,198 @@ handle_device_get_region_info(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
+    size_t i = 0;
 
     assert(vfu_ctx != NULL);
     assert(msg != NULL);
     assert(msg->out_fds == NULL);
 
-    if (msg->in_size != 16) {
+    if (msg->in_size != sizeof(vfio_user_region_io_fds_request_t)) {
         return ERROR_INT(EINVAL);
     }
 
-    uint max_sent_sub_regions = 0;
-    vfu_reg_info_t *vfu_reg = NULL;
-    reply_sub_region_t *rep = NULL;
-    sub_region_ioeventfd_t *ioefd = NULL;
-    region_io_fds_request_t *req = msg->in_data;
-    ioeventfd_list_t *sub_reg = NULL;
-    size_t nr_sub_reg = 0;
+    req = msg->in_data;
 
-    if (req->flags != 0 || req->count != 0 || msg->in_size < 16) {
+    if (req->flags != 0 || req->count != 0) {
         return ERROR_INT(EINVAL);
     }
 
     if (req->index >= vfu_ctx->nr_regions) {
-        vfu_log(vfu_ctx, LOG_DEBUG, "bad region index %d in get region info",
-                req->index);
+        vfu_log(vfu_ctx, LOG_DEBUG, "bad region index %d in get region io fds "
+                "info", req->index);
         return ERROR_INT(EINVAL);
     }
 
     vfu_reg = &vfu_ctx->reg_info[req->index];
-    sub_reg = LIST_FIRST(&vfu_reg->sub_reg_l);
-    while (sub_reg != NULL){
+    LIST_FOREACH(sub_reg, &vfu_reg->subregions, entry) {
         nr_sub_reg++;
-        sub_reg =  LIST_NEXT(sub_reg, pointers);
     }
 
-    if (req->argsz < 16 || req->argsz > SERVER_MAX_DATA_XFER_SIZE) {
+    if (req->argsz < sizeof(vfio_user_region_io_fds_reply_t) ||
+        req->argsz > SERVER_MAX_DATA_XFER_SIZE) {
         return ERROR_INT(EINVAL);
     }
 
-    max_sent_sub_regions = MIN((req->argsz - 16) / (int) sizeof(sub_region_ioeventfd_t), nr_sub_reg);
-    msg->out_size = 16 + max_sent_sub_regions * sizeof(sub_region_ioeventfd_t);
-    msg->out_data = calloc(1, msg->out_size);
-    rep = msg->out_data;
-    rep->index = req->index;
-    rep->count = max_sent_sub_regions;
-    rep->flags = 0;
-    rep->argsz = msg->out_size;
-
-    // Cannot send invalid files descriptors as it will cause the recv to fail and block
-    // No recovery is possible from that state in the test suite.
+    max_sent_sub_regions = MIN((req->argsz -
+                                sizeof(vfio_user_region_io_fds_reply_t)) /
+                                sizeof(vfio_user_sub_region_ioeventfd_t),
+                                nr_sub_reg);
+    msg->out_size = sizeof(vfio_user_region_io_fds_reply_t) +
+                    max_sent_sub_regions *
+                    sizeof(vfio_user_sub_region_ioeventfd_t);
+    msg->out_data = calloc(1, sizeof(vfio_user_region_io_fds_reply_t) +
+                           max_sent_sub_regions *
+                           sizeof(vfio_user_sub_region_ioeventfd_t));
+    if (msg->out_data == NULL) {
+        return -1;
+    }
+    reply = msg->out_data;
+    reply->index = req->index;
+    reply->count = max_sent_sub_regions;
+    reply->flags = 0;
+    reply->argsz = sizeof(vfio_user_region_io_fds_reply_t) +
+                          max_sent_sub_regions *
```

Comment:
still doesn't look right, this should be nr_sub_reg, not what we actually send

Reply:
Double checked on our spec doc against the full one in kvm. Our spec doc is missing information about the behavior of the system as it contains explicit reference to the behavior you are referencing here along with needing a change to be behavior when dealing with requests where the buffer is too small. Will start on these required changes and add the paragraphs to our spec doc on the repo.

Reply:
Are you sure there's something missing? When you say "the full one in kvm" - what are you talking about exactly? KVM knows nothing about vfio-user.

This is where in the spec this behaviour is documented:

```
argsz is the size required for the full reply payload (region info structure plus the size of any region capabilities)
```

plus a longer explanation in https://github.com/nutanix/libvfio-user/blob/master/docs/vfio-user.rst#message-sizes

Reply:
Sorry for the previous comment, I had found archives of the mailing lists with the draft proposal that had 2 additional paragraphs relating to the explicit behavior of argsz which appear to have been removed at some point and is instead explained in message size.


From the draft version of the spec
```
> The client sets the ``argsz`` field to indicate the maximum size of the response
> that the server can send, which must be at least the size of the response header
> plus space for the subregion array. If the full response size exceeds ``argsz``,
> then the server must respond only with the response header and the Region IO FD
> info structure, setting in ``argsz`` the buffer size required to store the full
> response. In this case, no file descriptors are passed back.  The client then
> retries the operation with a larger receive buffer.
```